ethcore/src/client/client.rs

```diff
@@ -1438,15 +1443,38 @@ impl RegistryInfo for Client {
 
 impl CallContract for Client {
 	fn call_contract(&self, block_id: BlockId, address: Address, data: Bytes) -> Result<Bytes, String> {
+		self.call_contract_from_sender(block_id, address, Address::default(), data)
+	}
+
+	fn call_contract_from_sender(&self, block_id: BlockId, address: Address, sender: Address, data: Bytes) -> Result<Bytes, String> {
 		let state_pruned = || CallError::StatePruned.to_string();
 		let state = &mut self.state_at(block_id).ok_or_else(&state_pruned)?;
 		let header = self.block_header_decoded(block_id).ok_or_else(&state_pruned)?;
 
-		let transaction = self.contract_call_tx(block_id, address, data);
-
-		self.call(&transaction, Default::default(), state, &header)
-			.map_err(|e| format!("{:?}", e))
-			.map(|executed| executed.output)
+		let transaction = self.contract_call_tx_from_sender(block_id, address, sender, data);
+
+		let executed = self.call(&transaction, Default::default(), state, &header).map_err(|e| format!("{:?}", e))?;
+		if let Some(exception) = executed.exception {
+			let mut error = format!("{:?}", exception);
+			if exception == VmError::Reverted {
+				let output = executed.output.clone();
+				fn format_unknown_output(output: &Vec<u8>) -> String {format!("{}", Token::Bytes(output.to_vec()))};
+				// Try to decode Solidity revert string. If not successful - return bytes
+				let msg = if output.len() > 4 {
+					let (error_selector, enc_string) = output.split_at(4);
+					if error_selector == [0x08, 0xc3, 0x79, 0xa0] {  // Error(string) selector
```

Comment:
This looks very much like a magic number if you don't know what the relationship to Solidity is. I'd add an explanation and/or a link to [this](https://solidity.readthedocs.io/en/v0.5.7/control-structures.html#error-handling-assert-require-revert-and-exceptions) in order to clarify what's going on here.

Reply:
Done :)